assets/src/edit-story/app/media/utils/getResourceFromMedia3p.js

```diff
@@ -54,81 +66,155 @@ import createResource from './createResource';
-      );
-    }
+function getImageUrls(m) {
+  if (m.type.toLowerCase() !== 'image') {
+    throw new Error('Invalid media type.');
+  }
 
-    const sizesFromBiggest = m.imageUrls
-      .sort((el1, el2) => el2.width - el1.width)
-      .map((u) => ({
-        file: m.name,
-        source_url: u.url,
-        mime_type: u.mimeType,
-        width: u.width,
-        height: u.height,
-      }));
-    const namedSizes = [
-      ['full', sizesFromBiggest[0]],
-      ['large', sizesFromBiggest[1]],
-      ...sizesFromBiggest
-        .slice(2, sizesFromBiggest.length - 1)
-        .map((u) => [u.width + '_' + u.height, u]),
-      ['web_stories_thumbnail', sizesFromBiggest[sizesFromBiggest.length - 1]],
-    ];
-    return Object.fromEntries(new Map(namedSizes));
+  // The rest of the application expects 3 named "sizes": "full", "large" and
+  // "web_stories_thumbnail". We use the biggest as "full", the next biggest
+  // as "large", and the smallest as "web_stories_thumbnail". The rest are
+  // named according to their size.
+  if (m.imageUrls?.length < 3) {
+    throw new Error('Invalid number of urls for asset. Need at least 3: ' + m);
   }
-  throw new Error('Invalid media type.');
+
+  const sizesFromBiggest = m.imageUrls
+    .sort((x, y) => y.width - x.width)
+    .map((u) => ({
+      file: m.name,
+      source_url: u.url,
+      mime_type: u.mimeType,
+      width: u.width,
+      height: u.height,
+    }));
+  const namedSizes = [
+    ['full', sizesFromBiggest[0]],
+    ['large', sizesFromBiggest[1]],
+    ...sizesFromBiggest
+      .slice(2, sizesFromBiggest.length - 1)
+      .map((u) => [u.width + '_' + u.height, u]),
+    ['web_stories_thumbnail', sizesFromBiggest[sizesFromBiggest.length - 1]],
+  ];
+  return Object.fromEntries(new Map(namedSizes));
 }
 
 /**
- * Generates a resource object from a Media3P object from the API.
+ * Converts the video urls in a Media3P Media object to the "sizes"
+ * object required for a Resource.
  *
- * @param {Media3pMedia} m A Media3P Media object.
- * @return {import('./createResource.js').Resource} Resource object.
+ * @param {Media3pMedia} m The Media3P Media object.
+ * @return {Object} The array of "sizes"-type objects.
  */
-function getResourceFromMedia3p(m) {
-  const urls = getUrls(m);
-  return createResource({
-    type: m.type.toLowerCase(),
-    mimeType: urls.full.mime_type,
-    creationDate: m.createTime,
-    src: urls.full.source_url,
-    width: urls.full.width,
-    height: urls.full.height,
-    poster: null, // TODO: Implement for videos.
-    posterId: null, // TODO: Implement for videos.
-    length: null, // TODO: Implement for videos.
-    lengthFormatted: null, // TODO: Implement for videos.
-    title: m.description,
-    alt: null,
-    local: false, // TODO: How does this interact with the rest?
-    sizes: urls, // TODO: Map with expected keys for canvas.
-    attribution: (m.author || m.registerUsageUrl) && {
+function getVideoUrls(m) {
```

Comment:
[optional] feels like there is a lot of shared logic between getVideoUrl and getImageUrl. Maybe it could be combined with a helper function?

Reply:
I started doing that, but there seems to be enough differences to warrant a divergence.
Perhaps this part can be extracted to a method though:

    .map((u) => ({
      file: m.name,
      source_url: u.url,
      mime_type: u.mimeType,
      width: u.width,
      height: u.height,
    }));

I'll update this in a follow-up PR.